From 4f0846108aa2c14126ebfbcc6b2e8b3150cdd5b2 Mon Sep 17 00:00:00 2001
From: chenj <chenj@lemote.com>
Date: Wed, 6 Nov 2013 15:19:04 +0800
Subject: [PATCH 040/130] Add ioremap.h for loongson platform

Add loongson specific plat_ioremap to utilize uncached acceleration
feature.

Change-Id: I5fb4cd821f8d9b15d308b82909cc58e42749bd48
Signed-off-by: Huacai Chen <chenhc@lemote.com>
Signed-off-by: Hongliang Tao <taohl@lemote.com>
Signed-off-by: Zhang Shuangshuang <zhangshuangshuang@ict.ac.cn>
---
 arch/mips/include/asm/mach-loongson64/ioremap.h | 44 +++++++++++++++++++++++++
 1 file changed, 44 insertions(+)
 create mode 100644 arch/mips/include/asm/mach-loongson64/ioremap.h

diff --git a/arch/mips/include/asm/mach-loongson64/ioremap.h b/arch/mips/include/asm/mach-loongson64/ioremap.h
new file mode 100644
index 00000000..367aca10
--- /dev/null
+++ b/arch/mips/include/asm/mach-loongson64/ioremap.h
@@ -0,0 +1,44 @@
+/*
+ *	include/asm-mips/mach-loongson64/ioremap.h
+ *
+ *	This program is free software; you can redistribute it and/or
+ *	modify it under the terms of the GNU General Public License
+ *	as published by the Free Software Foundation; either version
+ *	2 of the License, or (at your option) any later version.
+ */
+#ifndef __ASM_MACH_LOONGSON_IOREMAP_H
+#define __ASM_MACH_LOONGSON_IOREMAP_H
+
+#include <linux/types.h>
+
+/*
+ * Allow physical addresses to be fixed up to help peripherals located
+ * outside the low 32-bit range -- generic pass-through version.
+ */
+static inline phys_addr_t fixup_bigphys_addr(phys_addr_t phys_addr, phys_addr_t size)
+{
+	return phys_addr;
+}
+
+/* Add support for uncached accelerate */
+static inline void __iomem *plat_ioremap(phys_addr_t offset, unsigned long size,
+	unsigned long flags)
+{
+#if CONFIG_64BIT
+#define LOONGSON_UNCACHED_ACCEL_BASE _AC(0xb800000000000000, UL)
+	u64 base = LOONGSON_UNCACHED_ACCEL_BASE;
+
+	if (flags == _CACHE_UNCACHED_ACCELERATED &&
+	    offset + size < XKSEG - base) {
+		return (void __iomem *)(unsigned long) (base + offset);
+	}
+#endif
+	return NULL;
+}
+
+static inline int plat_iounmap(const volatile void __iomem *addr)
+{
+	return 0;
+}
+
+#endif /* __ASM_MACH_LOONGSON_IOREMAP_H */
-- 
2.11.0

